### DISCUSSION TITLE
Render pbf tiles

### DISCUSSION THREAD

**Discussion by p-lr** [maplibre/maplibre-native#2070] [Q&A]
I have a custom map renderer on Android which works with `Bitmap`s. It works really well with WMTS tile servers which provide png or jpeg tiles. However I'm having a hard time figuring out how to deal with vector tiles.
All I need is to convert pbf tiles to `Bitmap`s. Is it possible?

For example, given this pbf tile : https://basemaps.arcgis.com/arcgis/rest/services/World_Basemap_v2/VectorTileServer/tile/6/23/31.pbf

And this [style](https://cdn.arcgis.com/sharing/rest/content/items/b079fb8620f945cca0019db1cf611b45/resources/styles/root.json?f=json), is there a way to get `png`s or `jpeg`s from pbf tiles on Android?

**Comment by acalcutt** [comment]
[Tileserver-gl ](https://github.com/maptiler/tileserver-gl) is an option. it can take your pbf and style and serve rendered raster/wmts tiles rendered with maplibe-native

it doesn't do bitmap as far as i know though. it does jpg, png, and webp. bitmap isnt really a good web image type.

**Reply by p-lr** [reply]
Thanks, however Tileserver-gl is for server-side rendering (using node), while I need to do client-side rendering (on Android device).

**Reply by p-lr** [reply]
Also, converting to png or jpeg is perfectly fine for me. I'm editing my OP.

**Reply by acalcutt** [reply]
One thing I have tried before was using a combination of QGIS and tileserver-gl . In QGIS you use 
`Processing > Processing Toolbox (Toolbox) > Raster tools > Generate XYZ Tiles`

You can use tileserver-gl as a source and QGIS to pulls in the rendered tiles
![image (5)](https://github.com/maplibre/maplibre-native/assets/3792408/12693fe9-8fac-4111-a18e-d4e9873d92fd)
![image (6)](https://github.com/maplibre/maplibre-native/assets/3792408/68a1d734-110e-479c-8279-9bef441036f2)

**Reply by acalcutt** [reply]
It sounds like this isn't what you are looking for either though.  I'm guessing you would need to use the android sdk to do client side rendering on android, but I'm not too familiar with if it is able to render to images directly.

**Comment by acalcutt** [comment]
This is also an option,  It can make pre-rendered mbtiles from your styles
https://github.com/ConservationMetrics/mapgl-tile-renderer/

(although again is probably not what OP is looking for)
